File: jybot/executor.py
"""
jybot.executor -- order placement (paper + live).

Order rules (per spec)
----------------------
* Fixed-share sizing: every order uses ``size = FIXED_SHARES`` (BUY) or
  ``size <= position`` (SELL). We NEVER pass a USD ``amount`` as the share
  count -- share count and notional are kept distinct.
* BUY price  = best_ask * (1 + SLIPPAGE_BPS)         (rounded to tick)
* SELL price = best_bid * (1 - SLIPPAGE_BPS)         (rounded to tick)
* ORDER_TYPE:
    FOK = fill-or-kill   -> must fill in full or it is cancelled
    FAK = fill-and-kill  -> partial fills allowed, remainder cancelled
    GTC = good-till-cancel -> remainder rests on the book
* DRY_RUN (default): no real order is signed. The fill is simulated against the
  live order book so paper P&L is realistic.

The live path imports ``py-clob-client`` lazily -- it is only needed when both
``DRY_RUN=false`` and ``LIVE_TRADING=true`` (see Config.can_trade_live).
"""
from __future__ import annotations

import time
from dataclasses import dataclass, field
from typing import List, Optional

from jybot.config import Config
from jybot.log import logger
from jybot.markets import OrderBook, Token


@dataclass
class OrderResult:
    status: str                 # FILLED | PARTIAL | CANCELLED | REJECTED | FAILED | RESTING
    side: str                   # BUY | SELL
    token_id: str
    requested_size: float
    filled_size: float
    limit_price: float
    avg_price: float
    order_type: str
    order_id: Optional[str] = None
    reason: str = ""
    dry_run: bool = True

    @property
    def is_filled(self) -> bool:
        return self.status in ("FILLED", "PARTIAL") and self.filled_size > 0


def _round_to_tick(price: float, tick: float, *, up: bool) -> float:
    if tick <= 0:
        tick = 0.01
    steps = price / tick
    rounded = (math_ceil(steps) if up else math_floor(steps)) * tick
    # clamp to valid CLOB band
    rounded = max(tick, min(1.0 - tick, rounded))
    return round(rounded, 4)


def math_ceil(x: float) -> int:
    import math
    return int(math.ceil(x - 1e-9))


def math_floor(x: float) -> int:
    import math
    return int(math.floor(x + 1e-9))


def buy_limit_price(book: OrderBook, slippage_frac: float) -> Optional[float]:
    if not book.best_ask:
        return None
    raw = book.best_ask.price * (1.0 + slippage_frac)
    return _round_to_tick(raw, book.tick_size, up=True)


def sell_limit_price(book: OrderBook, slippage_frac: float) -> Optional[float]:
    if not book.best_bid:
        return None
    raw = book.best_bid.price * (1.0 - slippage_frac)
    return _round_to_tick(raw, book.tick_size, up=False)


def _simulate_fill(book: OrderBook, side: str, size: float, limit_price: float):
    """Walk the book to compute fillable size & average price at/under (BUY) or
    at/over (SELL) the limit price. Returns (filled_size, avg_price)."""
    filled = 0.0
    notional = 0.0
    if side == "BUY":
        # consume asks from lowest price up
        for lvl in book.asks:  # ascending
            if lvl.price > limit_price + 1e-9:
                break
            take = min(size - filled, lvl.size)
            if take <= 0:
                break
            filled += take
            notional += take * lvl.price
            if filled >= size - 1e-9:
                break
    else:  # SELL -- consume bids from highest price down
        for lvl in reversed(book.bids):  # descending
            if lvl.price < limit_price - 1e-9:
                break
            take = min(size - filled, lvl.size)
            if take <= 0:
                break
            filled += take
            notional += take * lvl.price
            if filled >= size - 1e-9:
                break
    avg = (notional / filled) if filled > 0 else limit_price
    return filled, avg


class Executor:
    """Places orders. One instance per run; lazily connects the live client."""

    def __init__(self, cfg: Config):
        self.cfg = cfg
        self._client = None
        self._client_ready = False

    # ── live client (lazy) ────────────────────────────────────────────────
    def _ensure_live_client(self) -> bool:
        if self._client_ready:
            return self._client is not None
        self._client_ready = True
        try:
            from py_clob_client.client import ClobClient
        except Exception as exc:
            logger.error(f"[exec] py-clob-client not installed: {exc}")
            logger.error("[exec] install it for live trading: pip install py-clob-client")
            self._client = None
            return False
        try:
            funder = self.cfg.funder_address or None
            client = ClobClient(
                host=self.cfg.clob_api_url,
                key=self.cfg.private_key,
                chain_id=self.cfg.chain_id,
                signature_type=self.cfg.signature_type,
                funder=funder,
            )
            client.set_api_creds(
                # creds object; py-clob-client accepts kwargs via ApiCreds
                creds=_make_api_creds(self.cfg),
            )
            self._client = client
            logger.info("[exec] live ClobClient ready")
            return True
        except Exception as exc:
            logger.error(f"[exec] failed to init ClobClient: {exc}")
            self._client = None
            return False

    # ── public API ────────────────────────────────────────────────────────
    def place_buy(self, token: Token, book: OrderBook, size: float) -> OrderResult:
        price = buy_limit_price(book, self.cfg.slippage_frac)
        if price is None:
            return OrderResult(
                status="REJECTED", side="BUY", token_id=token.token_id,
                requested_size=size, filled_size=0.0, limit_price=0.0,
                avg_price=0.0, order_type=self.cfg.order_type,
                reason="empty order book (no ask)", dry_run=not self.cfg.can_trade_live,
            )
        return self._place("BUY", token, book, size, price)

    def place_sell(self, token: Token, book: OrderBook, size: float, position: float) -> OrderResult:
        # SELL size can never exceed the current position.
        size = min(size, position)
        if size <= 0:
            return OrderResult(
                status="REJECTED", side="SELL", token_id=token.token_id,
                requested_size=size, filled_size=0.0, limit_price=0.0,
                avg_price=0.0, order_type=self.cfg.order_type,
                reason="no position to sell", dry_run=not self.cfg.can_trade_live,
            )
        price = sell_limit_price(book, self.cfg.slippage_frac)
        if price is None:
            return OrderResult(
                status="REJECTED", side="SELL", token_id=token.token_id,
                requested_size=size, filled_size=0.0, limit_price=0.0,
                avg_price=0.0, order_type=self.cfg.order_type,
                reason="empty order book (no bid)", dry_run=not self.cfg.can_trade_live,
            )
        return self._place("SELL", token, book, size, price)

    # ── internal ──────────────────────────────────────────────────────────
    def _place(self, side: str, token: Token, book: OrderBook, size: float, price: float) -> OrderResult:
        if self.cfg.can_trade_live:
            return self._place_live(side, token, size, price)
        return self._place_paper(side, token, book, size, price)

    def _place_paper(self, side: str, token: Token, book: OrderBook, size: float, price: float) -> OrderResult:
        filled, avg = _simulate_fill(book, side, size, price)
        order_type = self.cfg.order_type
        status, reason = _resolve_sim_status(order_type, size, filled)
        if status in ("CANCELLED", "REJECTED"):
            filled, avg = 0.0, price
        logger.info(
            f"[paper] {side} {token.outcome} size={size} @ {price:.3f} "
            f"({order_type}) -> {status} filled={filled:.2f} avg={avg:.3f}"
        )
        return OrderResult(
            status=status, side=side, token_id=token.token_id,
            requested_size=size, filled_size=filled, limit_price=price,
            avg_price=avg, order_type=order_type,
            order_id=f"paper-{int(time.time()*1000)}", reason=reason, dry_run=True,
        )

    def _place_live(self, side: str, token: Token, size: float, price: float) -> OrderResult:
        if not self._ensure_live_client():
            return OrderResult(
                status="FAILED", side=side, token_id=token.token_id,
                requested_size=size, filled_size=0.0, limit_price=price,
                avg_price=0.0, order_type=self.cfg.order_type,
                reason="live client unavailable", dry_run=False,
            )
        try:
            from py_clob_client.clob_types import OrderArgs, OrderType
            from py_clob_client.order_builder.constants import BUY, SELL

            poly_side = BUY if side == "BUY" else SELL
            order_args = OrderArgs(
                token_id=token.token_id,
                price=float(price),
                size=float(size),
                side=poly_side,
                fee_rate_bps=0,
            )
            signed = self._client.create_order(order_args)
            ot_map = {
                "FOK": OrderType.FOK,
                "FAK": getattr(OrderType, "FAK", OrderType.GTC),
                "GTC": OrderType.GTC,
            }
            poly_ot = ot_map.get(self.cfg.order_type, OrderType.FOK)
            resp = self._client.post_order(signed, orderType=poly_ot)

            order_id = (resp or {}).get("orderID") or (resp or {}).get("orderId")
            success = bool(resp and (resp.get("success") or order_id))
            matched = float((resp or {}).get("makingAmount", 0) or 0)
            status = "FILLED" if success and matched else ("RESTING" if success else "FAILED")
            logger.info(f"[live] {side} {token.outcome} size={size} @ {price:.3f} -> {resp}")
            return OrderResult(
                status=status, side=side, token_id=token.token_id,
                requested_size=size, filled_size=size if status == "FILLED" else 0.0,
                limit_price=price, avg_price=price, order_type=self.cfg.order_type,
                order_id=order_id, reason=str(resp)[:200], dry_run=False,
            )
        except Exception as exc:
            logger.error(f"[live] order failed: {exc}")
            return OrderResult(
                status="FAILED", side=side, token_id=token.token_id,
                requested_size=size, filled_size=0.0, limit_price=price,
                avg_price=0.0, order_type=self.cfg.order_type,
                reason=str(exc)[:200], dry_run=False,
            )


def _resolve_sim_status(order_type: str, size: float, filled: float):
    full = filled >= size - 1e-9
    if order_type == "FOK":
        if full:
            return "FILLED", "fully filled"
        return "CANCELLED", "FOK: insufficient liquidity for full fill"
    if order_type == "FAK":
        if full:
            return "FILLED", "fully filled"
        if filled > 0:
            return "PARTIAL", "FAK: partial fill, remainder cancelled"
        return "CANCELLED", "FAK: nothing marketable"
    # GTC
    if full:
        return "FILLED", "fully filled"
    if filled > 0:
        return "PARTIAL", "GTC: partial fill, remainder rests"
    return "RESTING", "GTC: nothing marketable yet, order rests"


def _make_api_creds(cfg: Config):
    """Build py-clob-client ApiCreds object."""
    from py_clob_client.clob_types import ApiCreds

    return ApiCreds(
        api_key=cfg.api_key,
        api_secret=cfg.api_secret,
        api_passphrase=cfg.api_passphrase,
    )

File: rust/src/executor.rs
//! 下单执行器。
//!
//! * DRY-RUN（默认）：不签名、不下真实单。按实时盘口模拟成交，纸面 PnL 真实。
//! * LIVE：把订单交给 Python 助手 `scripts/place_order.py`（使用官方
//!   py-clob-client 完成 EIP-712 签名与提交）。行情/盘口判断全在 Rust 走 WS，
//!   仅"真实提交订单"这一步复用官方 SDK——符合需求。
//!
//! 下单规则（与 Python 版一致）：
//!   BUY  价 = best_ask × (1 + SLIPPAGE_BPS)，按 tick 向上取整
//!   SELL 价 = best_bid × (1 - SLIPPAGE_BPS)，按 tick 向下取整，size ≤ 持仓
//!   FOK 全成或撤 / FAK 部分成交后撤 / GTC 剩余挂单

use crate::clob::{dec_f64, OrderBook};
use crate::config::Config;
use serde_json::json;
use std::sync::Arc;
use tracing::{error, info};

pub struct OrderResult {
    pub status: String, // FILLED | PARTIAL | CANCELLED | REJECTED | FAILED | RESTING
    pub side: String,
    pub filled_size: f64,
    pub avg_price: f64,
    pub limit_price: f64,
    pub reason: String,
}

impl OrderResult {
    pub fn is_filled(&self) -> bool {
        matches!(self.status.as_str(), "FILLED" | "PARTIAL") && self.filled_size > 0.0
    }
    fn rejected(side: &str, reason: &str) -> Self {
        OrderResult {
            status: "REJECTED".into(),
            side: side.into(),
            filled_size: 0.0,
            avg_price: 0.0,
            limit_price: 0.0,
            reason: reason.into(),
        }
    }
}

fn round_tick(price: f64, tick: f64, up: bool) -> f64 {
    let tick = if tick <= 0.0 { 0.01 } else { tick };
    let steps = price / tick;
    let r = if up { steps.ceil() } else { steps.floor() } * tick;
    r.clamp(tick, 1.0 - tick)
}

pub fn buy_limit_price(book: &OrderBook, slip: f64) -> Option<f64> {
    let ask = dec_f64(book.best_ask()?);
    Some(round_tick(ask * (1.0 + slip), dec_f64(book.tick_size), true))
}

pub fn sell_limit_price(book: &OrderBook, slip: f64) -> Option<f64> {
    let bid = dec_f64(book.best_bid()?);
    Some(round_tick(bid * (1.0 - slip), dec_f64(book.tick_size), false))
}

/// 沿盘口逐档撮合，返回 (成交量, 均价)。
fn simulate_fill(book: &OrderBook, side: &str, size: f64, limit: f64) -> (f64, f64) {
    let mut filled = 0.0;
    let mut notional = 0.0;
    if side == "BUY" {
        for (p, s) in &book.asks {
            let (p, s) = (dec_f64(*p), dec_f64(*s));
            if p > limit + 1e-9 {
                break;
            }
            let take = (size - filled).min(s);
            if take <= 0.0 {
                break;
            }
            filled += take;
            notional += take * p;
            if filled >= size - 1e-9 {
                break;
            }
        }
    } else {
        for (p, s) in &book.bids {
            let (p, s) = (dec_f64(*p), dec_f64(*s));
            if p < limit - 1e-9 {
                break;
            }
            let take = (size - filled).min(s);
            if take <= 0.0 {
                break;
            }
            filled += take;
            notional += take * p;
            if filled >= size - 1e-9 {
                break;
            }
        }
    }
    let avg = if filled > 0.0 { notional / filled } else { limit };
    (filled, avg)
}

fn resolve_sim_status(order_type: &str, size: f64, filled: f64) -> (&'static str, &'static str) {
    let full = filled >= size - 1e-9;
    match order_type {
        "FOK" => {
            if full {
                ("FILLED", "fully filled")
            } else {
                ("CANCELLED", "FOK: insufficient liquidity")
            }
        }
        "FAK" => {
            if full {
                ("FILLED", "fully filled")
            } else if filled > 0.0 {
                ("PARTIAL", "FAK: partial fill, remainder cancelled")
            } else {
                ("CANCELLED", "FAK: nothing marketable")
            }
        }
        _ => {
            if full {
                ("FILLED", "fully filled")
            } else if filled > 0.0 {
                ("PARTIAL", "GTC: partial fill, remainder rests")
            } else {
                ("RESTING", "GTC: rests on book")
            }
        }
    }
}

pub struct Executor {
    cfg: Arc<Config>,
}

impl Executor {
    pub fn new(cfg: Arc<Config>) -> Self {
        Self { cfg }
    }

    pub async fn place_buy(&self, token_id: &str, outcome: &str, book: &OrderBook, size: f64) -> OrderResult {
        let Some(price) = buy_limit_price(book, self.cfg.slippage_frac()) else {
            return OrderResult::rejected("BUY", "empty book (no ask)");
        };
        self.place("BUY", token_id, outcome, book, size, price).await
    }

    pub async fn place_sell(
        &self,
        token_id: &str,
        outcome: &str,
        book: &OrderBook,
        size: f64,
        position: f64,
    ) -> OrderResult {
        let size = size.min(position); // SELL 不超持仓
        if size <= 0.0 {
            return OrderResult::rejected("SELL", "no position to sell");
        }
        let Some(price) = sell_limit_price(book, self.cfg.slippage_frac()) else {
            return OrderResult::rejected("SELL", "empty book (no bid)");
        };
        self.place("SELL", token_id, outcome, book, size, price).await
    }

    async fn place(
        &self,
        side: &str,
        token_id: &str,
        outcome: &str,
        book: &OrderBook,
        size: f64,
        price: f64,
    ) -> OrderResult {
        if self.cfg.can_trade_live() {
            self.place_live(side, token_id, size, price).await
        } else {
            self.place_paper(side, outcome, book, size, price)
        }
    }

    fn place_paper(&self, side: &str, outcome: &str, book: &OrderBook, size: f64, price: f64) -> OrderResult {
        let (mut filled, mut avg) = simulate_fill(book, side, size, price);
        let (status, reason) = resolve_sim_status(&self.cfg.order_type, size, filled);
        if matches!(status, "CANCELLED" | "REJECTED") {
            filled = 0.0;
            avg = price;
        }
        info!(
            "[paper] {} {} size={} @ {:.3} ({}) -> {} filled={:.2} avg={:.3}",
            side, outcome, size, price, self.cfg.order_type, status, filled, avg
        );
        OrderResult {
            status: status.into(),
            side: side.into(),
            filled_size: filled,
            avg_price: avg,
            limit_price: price,
            reason: reason.into(),
        }
    }

    /// LIVE：调用 Python 官方 SDK 助手下单。
    async fn place_live(&self, side: &str, token_id: &str, size: f64, price: f64) -> OrderResult {
        let payload = json!({
            "token_id": token_id,
            "side": side,
            "size": size,
            "price": price,
            "order_type": self.cfg.order_type,
        });
        let mut cmd = tokio::process::Command::new(&self.cfg.python_bin);
        cmd.arg(&self.cfg.place_order_script);
        cmd.stdin(std::process::Stdio::piped());
        cmd.stdout(std::process::Stdio::piped());
        cmd.stderr(std::process::Stdio::piped());

        let child = match cmd.spawn() {
            Ok(c) => c,
            Err(e) => {
                error!("[live] 无法启动下单助手: {e}");
                return OrderResult {
                    status: "FAILED".into(),
                    side: side.into(),
                    filled_size: 0.0,
                    avg_price: 0.0,
                    limit_price: price,
                    reason: format!("spawn place_order.py failed: {e}"),
                };
            }
        };

        use tokio::io::AsyncWriteExt;
        let mut child = child;
        if let Some(mut stdin) = child.stdin.take() {
            let _ = stdin.write_all(payload.to_string().as_bytes()).await;
            let _ = stdin.shutdown().await;
        }
        let out = match child.wait_with_output().await {
            Ok(o) => o,
            Err(e) => {
                return OrderResult {
                    status: "FAILED".into(),
                    side: side.into(),
                    filled_size: 0.0,
                    avg_price: 0.0,
                    limit_price: price,
                    reason: format!("place_order.py error: {e}"),
                }
            }
        };
        let stdout = String::from_utf8_lossy(&out.stdout);
        info!("[live] {} {} size={} @ {:.3} -> {}", side, token_id, size, price, stdout.trim());

        let parsed: serde_json::Value = serde_json::from_str(stdout.trim()).unwrap_or(json!({}));
        let status = parsed.get("status").and_then(|v| v.as_str()).unwrap_or("FAILED");
        let filled = parsed.get("filled_size").and_then(|v| v.as_f64()).unwrap_or(0.0);
        let avg = parsed.get("avg_price").and_then(|v| v.as_f64()).unwrap_or(price);
        OrderResult {
            status: status.to_string(),
            side: side.into(),
            filled_size: filled,
            avg_price: avg,
            limit_price: price,
            reason: parsed.get("reason").and_then(|v| v.as_str()).unwrap_or("").to_string(),
        }
    }
}
